kong/runloop/wasm.lua

```diff
@@ -0,0 +1,646 @@
+      end
+
+      new_ref.hash = new_chain_hash
+      new_ref.c_plan = c_plan
+    end
+
+    store_chain_ref(state, new_ref)
+  end
+
+  return state
+end
+
+
+---
+-- Replace the current filter chain state with a new one.
+--
+-- This function does not do any I/O or other yielding operations.
+--
+---@param new kong.runloop.wasm.state
+local function set_state(new)
+  if type(new) ~= "table" then
+    error("bad argument #1 to 'set_state' (table expected, got " ..
+          type(new) .. ")", 2)
+  end
+
+  local old = STATE
+
+  if old.version == new.version then
+    log(DEBUG, "called with new version that is identical to the last")
+  end
+
+  STATE = new
+end
+
+
+---
+-- Conditionally rebuild and update the filter chain state.
+--
+-- If the current state matches the desired version, no update
+-- will be performed.
+--
+---@param  new_version? string
+---@return boolean?     ok
+---@return string?      error
+local function update_in_place(new_version)
+  if not ENABLED then
+    return true
+  end
+
+  new_version = new_version or get_version()
+  local old = STATE
+
+  if new_version == old.version then
+    log(DEBUG, "filter chain state is already up-to-date, no changes needed")
+    return true
+  end
+
+  local new, err = rebuild_state(kong.db, new_version, old)
+  if not new then
+    log(ERR, "failed rebuilding filter chain state: ", err)
+    return nil, err
+  end
+
+  set_state(new)
+
+  return true
+end
+
+
+
+---@param route?    { id: string }
+---@param service?  { id: string }
+---@return kong.runloop.wasm.filter_chain_reference?
+local function get_filter_chain_for_request(route, service)
+  local service_id = service and service.id
+  local route_id = route and route.id
+  local state = STATE
+
+  return get_chain_ref(state, TYPE_COMBINED, service_id, route_id)
+      or get_chain_ref(state, TYPE_SERVICE, service_id)
+      or get_chain_ref(state, TYPE_ROUTE, nil, route_id)
+end
+
+
+_M.get_version = get_version
+
+_M.update_in_place = update_in_place
+
+_M.set_state = set_state
+
+
+---@param kong_config table
+function _M.init(kong_config)
+  if not kong_config.wasm then
+    return
+  end
+
+  local modules = kong_config.wasm_modules_parsed
+  if not modules or #modules == 0 then
+    return
+  end
+
+  -- setup a DNS client for ngx_wasm_module
+  _G.dns_client = dns(kong_config)
+
+  proxy_wasm = require "resty.http.proxy_wasm"
+  ENABLED = true
+
+  ATTACH_OPTS.isolation = proxy_wasm.isolations.FILTER
+end
+
+
+---@return boolean? ok
+---@return string?  error
+function _M.init_worker()
+  if not ENABLED then
+    return true
+  end
+
+  local ok, err = update_in_place()
+  if not ok then
+    return nil, err
+  end
+
+  return true
+end
+
+
+local function set_proxy_wasm_property(property, value)
+  if not value then
+    return
+  end
+
+  local ok, err = proxy_wasm.set_property(property, value)
+  if not ok then
+    log(ERR, "failed to set ", property, " property: ", err)
```

Comment:
It could be good to either refer to "wasm" here as well, and perhaps en-quote the property name like ngx_wasm_module does for all directives/properties/attributes etc that are reporting an error.